FAI-380 [Scorecard Editor] Output validation

kelvah opened the pull request:
JIRA ticket: https://issues.redhat.com/browse/FAI-380

This PR adds validation for the outputs section.

manstis commented on packages/pmml-editor/src/editor/components/Outputs/atoms/OutputLabels.tsx:
```diff
@@ -29,12 +32,29 @@ interface OutputLabelsProps {
     <>
       {optype && OutputFieldLabel("OpType", optype)}
-      {targetField && OutputFieldLabel("TargetField", targetField)}
+      {targetFieldValidation.length > 0 ? (
```
What happens if it's not missing but either (a) references a non-existent `MininingField`, (b) does not need to be populated as there is only one `MiningField` with usage of `target` (but the `OutputField.targetField` is set in the PMML)?

kelvah replied:
fixed now. if the reference is non-existent it's automatically deleted.
if the `targetField` is populated but not required, it's still kept. it could be useful if it becomes required again.

manstis commented on packages/pmml-editor/src/editor/components/Outputs/atoms/OutputLabelsEditMode.tsx:
```diff
@@ -76,13 +80,20 @@ export const OutputLabelsEditMode = (props: OutputLabelsEditModeProps) => {
+      {targetFieldValidation.length > 0 ? (
+        <ValidationIndicatorLabel validations={targetFieldValidation} cssClass="output-fields-list__item__label">
+          <strong>TargetField:</strong>&nbsp;
+          <em>Missing</em>
```
Same comment as above.

kelvah replied:
Fixed as above. (The fix is in the scorecard reducer though)

manstis commented on packages/pmml-editor/src/editor/reducers/ScorecardReducer.ts:
```diff
@@ -70,6 +73,117 @@ export const ScorecardReducer: HistoryAwareReducer<Scorecard, ScorecardActions>
+        if (isOutputsTargetFieldRequired(state.MiningSchema.MiningField)) {
+          const startIndex = state.Output?.OutputField.length ?? 0;
+          action.payload.outputFields.forEach((name, index) => {
+            validation.set(
```
We traditionally (by accident really!) have not added validation messages to the registry in reducers but delegated to the "validation" components. Did you want to retain the approach for consistency? Although I suspect at least some of this type of code (validation on actions) could change when we move to validating after commit. I'll leave the decision up to you (just let me know what you decide so I don't pick it up as a review topic again!).

kelvah replied:
I agree with you. Validation should stay inside "validation" methods. I did it here directly in the action because it seemed the easiest way to do it without having to manually update the state with the new output fields.
As you pointed, this part will become useless and removed when moving the validation after commits, because we will validate the updated state.
If it's OK for you, I would leave it like that for now.

manstis reviewed (commented):
I've put a couple of comments on the code.

I also observed the following unexpected behaviours:

- Disable Rank Order when Value entered? (Rank Textbox is already disabled)
- Delete entered Value from about. Rank Textbox does not re-enable. It should.

- Set more than one MF field as "target". Outputs light up. Good.
- Correct MFs to pick a target field. Warnings disappear. Good.
- Go back to MF and change field back to "active" (from "target").
- Outputs now reference a MF field that is not type "target". Would expect this to be an error?

- Output Label shows a Target field although the field has been changed back to "active". Should Reducer clear OF.targetField if a MF is changed to not be "target"? 
- The Outputs Properties Edit screen does not populate the Target Field dropdown. Correct. I suspect this is by chance as we filter MF's based on usage (of "target"). 
- The PMML still shows it (as the field changed back to "active").
- These errors could be fixed if the Reducer cleared OF.targetField when MF.usage changes.

I've attached some files to the JIRA that may help you reproduce these.

kelvah commented:
> I've put a couple of comments on the code.
> 
> [...]

Thank you @manstis. 

I fixed the issue with the disabled rank field (and now I am disabling rank order too).

I also added the logic that automatically cleans the `targetField` property if it contains the name of a mining schema field that got deleted or that is no more of type `target`.

manstis reviewed (approved):
LGTM. Thank-you @kelvah.

manstis commented:
Hi @tiagobento can we get this one in please?

tiagobento commented on packages/pmml-editor/src/editor/components/Outputs/organisms/OutputFieldExtendedProperties.tsx:
```diff
@@ -224,9 +305,20 @@ export const OutputFieldExtendedProperties = (props: OutputFieldExtendedProperti
   );
 };
+
+const GenericSelectorEditor = (
+  id: string,
+  items: Array<string | GenericSelectorOption>,
+  selection: string,
+  onSelect: (_selection: string) => void,
+  isDisabled?: boolean
+) => {
+  return <GenericSelector id={id} items={items} selection={selection} onSelect={onSelect} isDisabled={isDisabled} />;
+};
```
Does that add something to `GenericSelector`? Why not use `GenericSelector` directly?

kelvah replied:
`GenericSelectorEditor` precedes my work here. I just moved it to the bottom of the file to have `OutputFieldExtendedProperties` interface and declaration on top.

I agree we could get rid of it but I didn't want to refactor the whole component in this PR. We have a [JIRA](https://issues.redhat.com/browse/FAI-385) for replacing all `Select` usages with `FormSelect`, I think we could take the chance to do some cleanup also.

tiagobento commented on packages/pmml-editor/src/editor/components/Outputs/molecules/OutputFieldEditRow.tsx:
```diff
@@ -126,7 +138,11 @@ const OutputFieldEditRow = (props: OutputFieldEditRowProps) => {
     }
   );
-
+  const { service } = useValidationService();
```
Since `useValidationService` has a `service` property, maybe this Custom Hook should become `useValidation` and you could change this to

```suggestion
  const validationService = useValidation().service;
```

WDYT?

manstis replied:
This has all changed in my PR #406.. 

It is now `ValidationRegistry` and the `string` literal paths changed to the builder I added.

I propose this PR be merged (once you're OK with it) and my #406 will then be updated to include the work in here.

I pasted a dependency chain of PMML PRs on `tooling` on Zulip and updated PR labels in GitHub.

kelvah replied:
:+1:

tiagobento commented on packages/pmml-editor/src/editor/reducers/ScorecardReducer.ts:
```diff
@@ -70,6 +73,160 @@ export const ScorecardReducer: HistoryAwareReducer<Scorecard, ScorecardActions>
+        break;
+
+      case Actions.DeleteMiningSchemaField:
+        if (state.MiningSchema.MiningField.length && state.Output?.OutputField.length) {
+          validation.clear(`models[${action.payload.modelIndex}].Output`);
+          state.Output.OutputField.forEach((outputField, outputFieldIndex) => {
+            if (outputField.targetField === action.payload.name) {
+              service.batch(state, `models[${action.payload.modelIndex}]`, draft => {
+                draft!.Output!.OutputField[outputFieldIndex] = {
+                  ...draft!.Output!.OutputField[outputFieldIndex],
+                  targetField: undefined
+                };
+                validateOutput(
+                  action.payload.modelIndex,
+                  { ...state.Output!.OutputField[outputFieldIndex], targetField: undefined },
+                  outputFieldIndex,
+                  state.MiningSchema.MiningField.filter(
+                    (miningField, miningFieldIndex) => miningFieldIndex !== action.payload.miningSchemaIndex
+                  ),
+                  validation
+                );
+              });
+            }
+          });
+          validateOutputs(
+            action.payload.modelIndex,
+            state.Output?.OutputField,
+            state.MiningSchema.MiningField.filter((field, index) => index !== action.payload.miningSchemaIndex),
+            validation
+          );
+        }
```
Lots of `!` here. Is there an alternative? What happens if `draft` is undefined?

manstis replied:
This too is set to be refactored once all the existing "validation" PRs are merged.

Then I'll be doing https://issues.redhat.com/browse/FAI-384 to remove much of the filering etc shown above.

tiagobento reviewed (commented):
Left some minor comments inline. Please consider them!
